### ISSUE TITLE
problem launching Concern

### ISSUE THREAD

**Issue by ovitus** [combatopera/Concern#30]
I'm getting this error message when trying to launch Concern, OS is Debian 10:

ovitus@pc:~$ Concern
Traceback (most recent call last):
  File "/home/ovitus/.local/bin/Concern", line 6, in <module>
    from concern.concern import main_Concern
  File "/home/ovitus/.local/lib/python2.7/site-packages/concern/concern.py", line 75
    import_module(f".consumer.{config.Concern.consumerName}", package = __package__).configure(config)
                                                           ^
SyntaxError: invalid syntax

**Comment by combatopera** [comment]
Hi @ovitus,

The problem here seems to be that Concern was installed as a Python 2 app, but it requires a reasonably recent release of Python 3. Can you share the install command you used, did you try pip3 as per [the readme](https://github.com/combatopera/Concern/blob/master/README.md)?

**Comment by ovitus** [comment]
I had it installed with both pip and pip3. I uninstalled both and ran through the installation again.

ovitus@pc:~$ Concern
Traceback (most recent call last):
  File "/home/ovitus/.local/bin/Concern", line 6, in <module>
    from concern.concern import main_Concern
  File "/home/ovitus/.local/lib/python3.7/site-packages/concern/concern.py", line 26, in <module>
    from screen import stuffablescreen
  File "/home/ovitus/.local/lib/python3.7/site-packages/screen.py", line 18, in <module>
    from lagoon import screen
ImportError: cannot import name 'screen' from 'lagoon' (/home/ovitus/.local/lib/python3.7/site-packages/lagoon/__init__.py)

I got that, installed screen on a whim, but now it shows this:

ovitus@pc:~$ Concern
Traceback (most recent call last):
  File "/home/ovitus/.local/lib/python3.7/site-packages/aridity/config.py", line 63, in __getattr__
    obj = self._context.resolved(*path) # TODO LATER: Guidance for how lazy non-scalars should be in this situation.
  File "/home/ovitus/.local/lib/python3.7/site-packages/aridity/context.py", line 60, in resolved
    return self._resolved(path, self._findresolvable(path), kwargs) if path else self
  File "/home/ovitus/.local/lib/python3.7/site-packages/aridity/context.py", line 92, in _findresolvable
    raise NoSuchPathException(path)
aridity.util.NoSuchPathException: ('printf',)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/ovitus/.local/bin/Concern", line 10, in <module>
    sys.exit(main_Concern())
  File "/home/ovitus/.local/lib/python3.7/site-packages/concern/concern.py", line 54, in main_Concern
    config.printf(". %s", resource_filename(__name__, 'Concern.arid'))
  File "/home/ovitus/.local/lib/python3.7/site-packages/aridity/config.py", line 65, in __getattr__
    raise AttributeError(name) # XXX: Misleading?
AttributeError: printf

**Comment by combatopera** [comment]
Cool, that's good progress. I guess pip(2) and pip3 install the Concern script in the same place, so whichever you ran last would win. The current error you're getting is because I did some releases today and I'm sloppy with declaring dependency versions, so you still have older dependencies installed from the last attempt and pip3 feels no need to upgrade them. Can you try running the install command again with the --upgrade option? You shouldn't need to uninstall first.

**Comment by ovitus** [comment]
Hey, tried to upgrade, but I think it's already current:

ovitus@pc:~$ pip3 install --upgrade --user Concern
Requirement already up-to-date: Concern in ./.local/lib/python3.7/site-packages (16)
Requirement already satisfied, skipping upgrade: timelyOSC in ./.local/lib/python3.7/site-packages (from Concern) (3)
Requirement already satisfied, skipping upgrade: aridity in ./.local/lib/python3.7/site-packages (from Concern) (32)
Requirement already satisfied, skipping upgrade: lagoon in ./.local/lib/python3.7/site-packages (from Concern) (17)
Requirement already satisfied, skipping upgrade: pyparsing in ./.local/lib/python3.7/site-packages (from aridity->Concern) (2.4.7)
Requirement already satisfied, skipping upgrade: diapyr in ./.local/lib/python3.7/site-packages (from lagoon->Concern) (16)

Still getting the same error message.

**Comment by combatopera** [comment]
Thanks for the output. The problem is aridity, which should be version 33. I've had issues before where the latest release doesn't get picked up, not sure why it happens, possibly PyPI being slow combined with pip3 version. Try forcing it like this: pip3 install --user aridity==33

**Comment by ovitus** [comment]
ovitus@pc:~/.local/bin$ ./Concern 
/usr/bin/bash: lc2jack: command not found
Traceback (most recent call last):
  File "./Concern", line 10, in <module>
    sys.exit(main_Concern())
  File "/home/ovitus/.local/lib/python3.7/site-packages/concern/concern.py", line 71, in main_Concern
    import_module(f".consumer.{config.Concern.consumerName}", package = __package__).configure(config)
  File "/home/ovitus/.local/lib/python3.7/site-packages/concern/consumer/pym2149.py", line 31, in configure
    consumerinfo, = _getconfig(config, 'OSC')
  File "/home/ovitus/.local/lib/python3.7/site-packages/concern/consumer/pym2149.py", line 26, in _getconfig
    values = list(map(eval, bash._c(command).splitlines()))
  File "/home/ovitus/.local/lib/python3.7/site-packages/lagoon/program.py", line 140, in __call__
    return xform(subprocess.run(cmd, **kwargs))
  File "/usr/lib/python3.7/subprocess.py", line 487, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['/usr/bin/bash', '-c', "lc2jack --config 'pianorollenabled = false' /dev/null --repr OSC"]' returned non-zero exit status 127.

I think lc2jack is part of your pym2149 package, I had issues installing that via pip3, these are the errors, let me know if you need the entire output:

  Failed building wheel for pym2149
  Failed building wheel for outjack
  Failed building wheel for PyAudio
Command "/usr/bin/python3 -u -c "import setuptools, tokenize;__file__='/tmp/pip-install-fkw7tw1j/outjack/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-record-v8tamdgj/install-record.txt --single-version-externally-managed --compile --user --prefix=" failed with error code 1 in /tmp/pip-install-fkw7tw1j/outjack/

**Comment by combatopera** [comment]
Yes, please post entire output. I don't know why it's attempting to build wheels, it doesn't on my system.

**Comment by combatopera** [comment]
Actually, I might know why - try installing "wheel" and see if that helps.

**Comment by ovitus** [comment]
Already tried installing "wheel" and "setuptools" libraries. Same error message occurs, I attached the entire output.

[pym2149_pip3.txt](https://github.com/combatopera/Concern/files/5059339/pym2149_pip3.txt)

**Comment by combatopera** [comment]
Thanks, before we install any header packages can you try upgrading to latest pip like this: pip3 install --user --upgrade pip

**Comment by ovitus** [comment]
Ok, I've attached output from the latest attempt at building pym2149 after upgrading pip.

[pym2149_pip3_2.txt](https://github.com/combatopera/Concern/files/5059524/pym2149_pip3_2.txt)

**Comment by combatopera** [comment]
Cool, thanks for persevering, this thread will really help me write better documentation. Two of the errors can be fixed by installing jackd and portaudio headers:
```
sudo apt-get install portaudio19-dev libjack-jackd2-dev
```
To fix the other error you'll need to get resid-0.16.tar.gz (SID chip emulator) from http://www.zimmers.net/anonftp/pub/cbm/crossplatform/emulators/resid/index.html then unpack it and do the following:
```
./configure --enable-shared && make && sudo make install
```

**Comment by ovitus** [comment]
I've gotten it working following your instruction. Thank you for this!

**Comment by ovitus** [comment]
It launches to the screen session now, I can print "Hello, World," but when I execute any commands from the tutorial (creating a Player object), I see an error in the log:

NameError: name 'Player' is not defined

I've attached complete text from the screen session:
[Concern_startup.txt](https://github.com/combatopera/Concern/files/5060508/Concern_startup.txt)
[Concern_test.txt](https://github.com/combatopera/Concern/files/5060509/Concern_test.txt)

**Comment by combatopera** [comment]
p1 and Player are FoxDot objects, they don't exist in pym2149. Do you want to make music with FoxDot or with pym2149?

To use FoxDot you'll need to remove the line of config in ~/.settings.arid , and complete FoxDot's own installation instructions https://foxdot.org/installation/

To use pym2149 there is no tutorial, just the demos at the bottom of the readme https://github.com/combatopera/Concern#pym2149

**Comment by ovitus** [comment]
I updated ~/.settings.arid and FoxDot is sending notes to SC now. Is there a shortcut for Clock.clear() ?

Thanks

**Comment by combatopera** [comment]
Cool, glad you've got it working. There aren't any shortcuts implemented at the moment, I've now raised #32 to fix that. Is there anything else you need regarding launching Concern?

**Comment by ovitus** [comment]
That's all I can see right now, you've been very helpful. You can close the ticket out if you want.
